feat(Usage Reports): add billing snapshot config service to usage reports

manukm-ibm opened the pull request:
## PR summary
Introducing new CRUD APIs to configure billing reports snapshot.
TRI: https://github.ibm.com/BSS/architecture/blob/master/design-docs/services/billing/apptio-integration.md
Issue: https://github.ibm.com/BSS/metering-deliverables/issues/1990

## PR Checklist
Please make sure that your PR fulfills the following requirements:  
- [x] The commit message follows the [Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## Current vs new behavior  
<!-- Please describe the current behavior that you are modifying and the new behavior. -->

## Does this PR introduce a breaking change?    
- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information
<!-- Please add any additional information that would help reviewers evaluate your PR -->

pyrooka committed:
feat: bump Go versions (#271)

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>
Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
Update version 0.45.0 -> 0.46.0 [skip ci]

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
chore(release): 0.46.0 release notes [skip ci]

# [0.46.0](https://github.com/IBM/platform-services-go-sdk/compare/v0.45.0...v0.46.0) (2023-08-18)

### Features

* bump Go versions ([#271](https://github.com/IBM/platform-services-go-sdk/issues/271)) ([d5e0092](https://github.com/IBM/platform-services-go-sdk/commit/d5e009266f93f130743de2a9e91d38219508d3c6))

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/IBM/platform-services-go-sdk?pullRequest=272) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/IBM/platform-services-go-sdk?pullRequest=272) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/IBM/platform-services-go-sdk?pullRequest=272) it.</sub>

padamstx commented on usagereportsv4/usage_reports_v4.go:
```diff
@@ -1029,6 +1512,90 @@ func (options *GetOrgUsageOptions) SetHeaders(param map[string]string) *GetOrgUs
+	Month *string `json:"month" validate:"required"`
+
+	// Timestamp in milliseconds for which billing report snapshot is requested.
+	DateFrom *float64 `json:"date_from,omitempty"`
```
this should probably be of type `*int64`.  Are you sure that the API definition is correct and agrees with the server implementation?

padamstx commented on usagereportsv4/usage_reports_v4.go:
```diff
@@ -2203,132 +2812,736 @@ func UnmarshalResourcesSummary(m map[string]json.RawMessage, result interface{})
+	Files []SnapshotListSnapshotsItemFilesItem `json:"files,omitempty"`
+
+	// Timestamp at which snapshot is captured.
+	ProcessedAt *float64 `json:"processed_at,omitempty"`
```
Another timestamp type property that is of type float64.

padamstx commented on usagereportsv4/usage_reports_v4.go:
```diff
@@ -2203,132 +2812,736 @@ func UnmarshalResourcesSummary(m map[string]json.RawMessage, result interface{})
+	CosEndpoint *string `json:"cos_endpoint,omitempty"`
+
+	// Timestamp in milliseconds when the snapshot configuration was created.
+	CreatedAt *float64 `json:"created_at,omitempty"`
```
another timestamp that is a float64 value

padamstx commented on usagereportsv4/usage_reports_v4_integration_test.go:
```diff
@@ -118,22 +96,15 @@ var _ = Describe(`UsageReportsV4 Integration Tests`, func() {
 			getAccountSummaryOptions := &usagereportsv4.GetAccountSummaryOptions{
-				AccountID:    &accountID,
-				Billingmonth: &billingMonth,
+				AccountID: core.StringPtr("testString"),
```
You have re-generated the integration tests and overwritten the code that was there previously, including manually developed code.  Is this really what you want?
You can see from this set of diffs that this test block previously used `accountId` and `billingMonth` which are variables that are obtained from the .env configuration file.

padamstx reviewed (changes requested):
There are a few questions/issues to be addressed:
1. Did you really intend to completely overwrite the previous version of the integration tests?   There was manually developed code there that probably should be retained.  Please see https://github.com/IBM/platform-services-go-sdk/blob/main/update_service.md#updating-integration-testsexamples for instructions on how to update the integration tests and examples code while still retaining the previous manual changes.
2. There were no changes to the examples code, but yet the updates to the generated service code show 5 new operations.
3. It looks like a handful of timestamp-related properties that are supposed to hold "millisecond" values were defined in the API definition as "type number", which results in a language-specific floating point type being used in the generated code.   Are you sure that the server behavior supports a floating point value for these properties?   I would expect the API definition to use "type integer" and for the fields to be defined with a language-specific integer type.

manukm-ibm committed:
feat(Usage Reports): add billing snapshot config service to usage reports

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

Rajesh-Pirati committed:
feat(IAM Policy Management): fix  template metadata as part of policies (#274)

Signed-off-by: Rajesh K Pirati <Rajesh.Pirati@ibm.com>
Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
Update version 0.46.0 -> 0.47.0 [skip ci]

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
chore(release): 0.47.0 release notes [skip ci]

# [0.47.0](https://github.com/IBM/platform-services-go-sdk/compare/v0.46.0...v0.47.0) (2023-08-28)

### Features

* **IAM Policy Management:** fix  template metadata as part of policies ([#274](https://github.com/IBM/platform-services-go-sdk/issues/274)) ([542046f](https://github.com/IBM/platform-services-go-sdk/commit/542046f552aa1cbbdd744d5549f9357020840ff5))

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

padamstx commented:
@manukm-ibm I took a fresh look at the changes and it looks like 6 operations have been added to the API (based on the changes to the generated service code):
- CreateReportsSnapshotConfig
- GetReportsSnapshotConfig
- UpdateReportsSnapshotConfig
- DeleteReportsSnapshotConfig
- GetReportsSnapshot

We'll need integration tests and working examples for each of those new operations.  It might help to generate the int tests and examples off to the side (i.e. into a different output directory), then copy/paste individual sections for each new operation into the checked-in version of the int tests and examples.
Reference: https://github.com/IBM/platform-services-go-sdk/blob/main/update_service.md#updating-integration-testsexamples

padamstx committed:
fix: enable builds on windows (#273)

This commit updates the project to enable
builds to be performed on Windows.
I also bumped the go core version to the current latest.

Signed-off-by: Phil Adams <phil_adams@us.ibm.com>
Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
chore(release): 0.47.1 release notes [skip ci]

## [0.47.1](https://github.com/IBM/platform-services-go-sdk/compare/v0.47.0...v0.47.1) (2023-08-31)

### Bug Fixes

* enable builds on windows ([#273](https://github.com/IBM/platform-services-go-sdk/issues/273)) ([b39d2d5](https://github.com/IBM/platform-services-go-sdk/commit/b39d2d53b2b58b8b2b02335f6f0d878b33e8e712))

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
Update version 0.47.0 -> 0.47.1 [skip ci]

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm committed:
feat (Usage Reports): Temporary change

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

apoorva9s14 committed:
feat(Resource Controller): support resource group filter for list reclamations (#264)

Co-authored-by: Phil Adams <phil_adams@us.ibm.com>
Signed-off-by: manu.k.m <manu.k.m@ibm.com>

padamstx commented on go.mod:
```diff
@@ -1,4 +1,4 @@
-module github.com/IBM/platform-services-go-sdk
+module github.com/manukm-ibm/platform-services-go-sdk
```
Please revert this change.  It looks like it's causing errors in the build.

manukm-ibm replied:
Done. This was a temporary change to test my TF scripts

semantic-release-bot committed:
chore(release): 0.48.0 release notes [skip ci]

# [0.48.0](https://github.com/IBM/platform-services-go-sdk/compare/v0.47.1...v0.48.0) (2023-09-01)

### Features

* **Resource Controller:** support resource group filter for list reclamations ([#264](https://github.com/IBM/platform-services-go-sdk/issues/264)) ([6985653](https://github.com/IBM/platform-services-go-sdk/commit/6985653a853b73e06189b40547f38afffa6513c0))

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
Update version 0.47.1 -> 0.48.0 [skip ci]

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm committed:
feat (Usage Reports): Revert temporary change

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm committed:
Temperory Change

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm committed:
feat(Usage Reports): Revert Temperory Change

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

hariarla committed:
fix(IAM Access Groups): rRemove update action control from assertions (#276)

Signed-off-by: HARI K ARLA <hariarla@in.ibm.com>
Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
chore(release): 0.48.1 release notes [skip ci]

## [0.48.1](https://github.com/IBM/platform-services-go-sdk/compare/v0.48.0...v0.48.1) (2023-09-05)

### Bug Fixes

* **IAM Access Groups:** rRemove update action control from assertions ([#276](https://github.com/IBM/platform-services-go-sdk/issues/276)) ([75ec5ee](https://github.com/IBM/platform-services-go-sdk/commit/75ec5ee4d6924c7a785c1349e05cfd5060fc94ae))

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
Update version 0.48.0 -> 0.48.1 [skip ci]

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

JonahFarc committed:
feat(Context-Based Restrictions): add service group support to operations (#277)

Signed-off-by: Dylan <Dylan.Yu@ibm.com>
Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
Update version 0.48.1 -> 0.49.0 [skip ci]

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

semantic-release-bot committed:
chore(release): 0.49.0 release notes [skip ci]

# [0.49.0](https://github.com/IBM/platform-services-go-sdk/compare/v0.48.1...v0.49.0) (2023-09-13)

### Features

* **Context-Based Restrictions:** add service group support to operations ([#277](https://github.com/IBM/platform-services-go-sdk/issues/277)) ([8ecdfa5](https://github.com/IBM/platform-services-go-sdk/commit/8ecdfa520c0254fbc366b80680b4bc7eab8565d5))

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm committed:
feat(Usage Reports): add integration tests and examples for billing snapshot config service

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm committed:
Merge branch 'IBM:main' into bill-snapshot

manukm-ibm committed:
feat(Usage Reports): add required config properties for examples run

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm committed:
Merge branch 'bill-snapshot' of https://github.com/manukm-ibm/platform-services-go-sdk into bill-snapshot

manukm-ibm committed:
feat(Usage Reports): revert copyright year change

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

padamstx commented on usagereportsv4/usage_reports_v4_integration_test.go:
```diff
@@ -355,4 +373,108 @@ var _ = Describe(`UsageReportsV4 Integration Tests`, func() {
+			shouldSkipTest()
+		})
+		It(`CreateReportsSnapshotConfig(createReportsSnapshotConfigOptions *CreateReportsSnapshotConfigOptions)`, func() {
+			fmt.Printf("TEST VALUES: %s", cosBucket)
+			fmt.Printf("TEST VALUES:%s", cosLocation)
```
```suggestion
			fmt.Fprintf(GinkgoWriter, "COS Bucket: %s", cosBucket)
			fmt.Fprintf(GinkgoWriter, "COS Location:%s", cosLocation)
```
Please use `fmt.Fprintf(GinkgoWriter,...)` instead of `fmt.Printf(...)`.  This will ensure that output is displayed only when the `-ginkgo.v` command line option is used.
You can find other instances of this in the same source file.

manukm-ibm replied:
Oh this was a printf for testing. Will remove it.

padamstx reviewed (changes requested):
Looks good, but I had just one minor suggested change in the integration tests.

padamstx commented:
@manukm-ibm For some reason, it looks like this PR contains 28 commits, but most of those are commits that already were merged into main, so this looks pretty strange :)   Could you perhaps rebase your feature branch on the current main branch from the primary repo and clear that up?

manukm-ibm committed:
feat(Usage Reports): minor changes to usage reports

Signed-off-by: manu.k.m <manu.k.m@ibm.com>

manukm-ibm commented:
Integration test results of new changes
<img width="791" alt="Screenshot 2023-09-14 at 11 59 49 AM" src="https://github.com/IBM/platform-services-go-sdk/assets/137261252/46d235d7-1568-43f4-a1e2-4c478f0a082d">

Examples results of new changes
<img width="1251" alt="Screenshot 2023-09-14 at 12 44 16 PM" src="https://github.com/IBM/platform-services-go-sdk/assets/137261252/f772a1e5-3db5-4f44-a3dc-803152f578ec">

padamstx reviewed (approved):
LGTM

ibm-devx-sdk commented:
:tada: This PR is included in version 0.50.0 :tada:

The release is available on [GitHub release](https://github.com/IBM/platform-services-go-sdk/releases/tag/v0.50.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
